File: packages/py-moose-lib/moose_lib/config/config_file.py
"""
Configuration file handling for Moose.

This module provides functionality for reading and parsing the moose.config.toml file,
which contains project-wide configuration settings.
"""
import os
import tomllib
from dataclasses import dataclass
from typing import Optional

@dataclass
class ClickHouseConfig:
    """ClickHouse configuration settings from moose.config.toml."""
    host: str
    host_port: int
    user: str
    password: str
    db_name: str
    use_ssl: bool = False
    native_port: Optional[int] = None

@dataclass
class ProjectConfig:
    """Project configuration from moose.config.toml."""
    language: str
    clickhouse_config: ClickHouseConfig

def find_config_file(start_dir: str = os.getcwd()) -> Optional[str]:
    """Find moose.config.toml by walking up directory tree.

    Args:
        start_dir: Directory to start searching from. Defaults to current directory.

    Returns:
        Path to moose.config.toml if found, None otherwise.
    """
    current_dir = os.path.abspath(start_dir)
    while True:
        config_path = os.path.join(current_dir, "moose.config.toml")
        if os.path.exists(config_path):
            return config_path

        parent_dir = os.path.dirname(current_dir)
        # Reached root directory
        if parent_dir == current_dir:
            break
        current_dir = parent_dir
    return None

def read_project_config() -> ProjectConfig:
    """Read and parse moose.config.toml.

    Returns:
        ProjectConfig object containing parsed configuration.
    """
    config_path = find_config_file()
    if not config_path:
        raise FileNotFoundError(
            "moose.config.toml not found in current directory or any parent directory"
        )

    try:
        with open(config_path, "rb") as f:
            config_data = tomllib.load(f)

        clickhouse_config = ClickHouseConfig(
            host=config_data["clickhouse_config"]["host"],
            host_port=config_data["clickhouse_config"]["host_port"],
            user=config_data["clickhouse_config"]["user"],
            password=config_data["clickhouse_config"]["password"],
            db_name=config_data["clickhouse_config"]["db_name"],
            use_ssl=config_data["clickhouse_config"].get("use_ssl", False),
            native_port=config_data["clickhouse_config"].get("native_port")
        )

        return ProjectConfig(
            language=config_data["language"],
            clickhouse_config=clickhouse_config
        )
    except Exception as e:
        raise RuntimeError(f"Failed to parse moose.config.toml: {e}")

File: packages/ts-moose-lib/src/config/configFile.ts
import path from "node:path";
import * as toml from "toml";

/**
 * ClickHouse configuration from moose.config.toml
 */
export interface ClickHouseConfig {
  host: string;
  host_port: number;
  user: string;
  password: string;
  db_name: string;
  use_ssl?: boolean;
  native_port?: number;
}

/**
 * Project configuration from moose.config.toml
 */
export interface ProjectConfig {
  language: string;
  clickhouse_config: ClickHouseConfig;
  // Add other config sections as needed
}

/**
 * Error thrown when configuration cannot be found or parsed
 */
export class ConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ConfigError";
  }
}

/**
 * Walks up the directory tree to find moose.config.toml
 */
async function findConfigFile(
  startDir: string = process.cwd(),
): Promise<string | null> {
  const fs = await import("node:fs");

  let currentDir = path.resolve(startDir);

  while (true) {
    const configPath = path.join(currentDir, "moose.config.toml");
    if (fs.existsSync(configPath)) {
      return configPath;
    }

    const parentDir = path.dirname(currentDir);
    if (parentDir === currentDir) {
      // Reached root directory
      break;
    }
    currentDir = parentDir;
  }

  return null;
}

/**
 * Reads and parses the project configuration from moose.config.toml
 */
export async function readProjectConfig(): Promise<ProjectConfig> {
  const fs = await import("node:fs");
  const configPath = await findConfigFile();
  if (!configPath) {
    throw new ConfigError(
      "moose.config.toml not found in current directory or any parent directory",
    );
  }

  try {
    const configContent = fs.readFileSync(configPath, "utf-8");
    const config = toml.parse(configContent) as ProjectConfig;
    return config;
  } catch (error) {
    throw new ConfigError(`Failed to parse moose.config.toml: ${error}`);
  }
}
